extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -629,6 +646,43 @@ export class Bind {
   }
 
   /**
+   *
+   */
+  beginObservingElementForMutations_(element) {
+    // TODO(kmh287): What if parent is the body tag?
+    // TODO(kmh287): Generify logic for node observation strategy
+    // when bind supprots more dynamic nodes.
+    const elementToObserve = element.parentElement;
+    this.mutationObserver_.observe(elementToObserve, {childList: true});
+  }
+
+  /**
+   * Respond to observed mutations. Meant to be used with MutationObserver
+   *
+   * @param mutations {Array<MutationRecord>}
+   */
+  onMutationsObserved_(mutations) {
+    mutations.forEach(mutation => {
+      const mutationPromises = [];
+      const removedNodes = mutation.removedNodes;
+      for (let i = 0; i < removedNodes.length; i++) {
+        const removedNode = removedNodes[i];
+        mutationPromises.push(this.removeBindingsForNode_(removedNode));
```

Comment:
We have a very subtle race condition here.

Say some expression `1` gets all its bound elements removed. But, we add a new node that uses the same `1` expression. So, we'll issue a `bind.removeBindingsWithExpressionStrings`, then a `bind.addBindings`. But, because web worker messages aren't ordered (they likely are, but not guaranteed), we could receive `bind.addBindings` then `bind.removeBindingsWithExpressionStrings`, causing us to lose the new `1` expression binding.

Instead, we could add the bindings first, then remove the old bindings. Because the removal will now account for the added ones, we won't lose our `1` expression.

Also, we'll probably be able to send smaller messages later on. 😀